### ISSUE TITLE
Differences in mypy output between pre-commit and VSCode

### ISSUE THREAD

**Issue by ckolbPTB** [PTB-MR/mrpro#47]
(@fzimmermann89 I moved this discussion to a new issue)

When I run 
`pre-commit run --all-files `
in the mrpro folder then the mypy checks all pass.

When I run 
`mypy src/mrpro `
(which is I think what VS Code is running) then I get:
_src/mrpro/data/_SpatialDimension.py:88: error: Incompatible return value type (got "SpatialDimension[int]", expected "SpatialDimension[Tensor]")  [return-value]
src/mrpro/data/_SpatialDimension.py:88: error: Incompatible return value type (got "SpatialDimension[float]", expected "SpatialDimension[Tensor]")  [return-value]
src/mrpro/data/_SpatialDimension.py:88: error: Argument 1 to "SpatialDimension" has incompatible type "Tensor"; expected "int"  [arg-type]
src/mrpro/data/_SpatialDimension.py:88: error: Argument 1 to "SpatialDimension" has incompatible type "Tensor"; expected "float"  [arg-type]
src/mrpro/data/_SpatialDimension.py:88: error: Argument 2 to "SpatialDimension" has incompatible type "Tensor"; expected "int"  [arg-type]
src/mrpro/data/_SpatialDimension.py:88: error: Argument 2 to "SpatialDimension" has incompatible type "Tensor"; expected "float"  [arg-type]
src/mrpro/data/_SpatialDimension.py:88: error: Argument 3 to "SpatialDimension" has incompatible type "Tensor"; expected "int"  [arg-type]
src/mrpro/data/_SpatialDimension.py:88: error: Argument 3 to "SpatialDimension" has incompatible type "Tensor"; expected "float"  [arg-type]
Found 8 errors in 1 file (checked 12 source files)_

Does anybody know what the pre-commit is doing differently with mypy?

**Comment by fzimmermann89** [comment]
I will take a look. My VScode does NOT complain.
Could you maybe post the output of 
* whereis mypy (in the shell that caused the output you posted)
* mypy -V
* the python version
* the mypy-type-checker.importStrategy vscode setting

It could be that different versions behave differently with regards to how smart they are in figuring out the types. I think the annotations are correct and sufficient:

The conversion is typed to always return a Tensor thus x y and z are Tensor. 
Thus we return SpatialDimension(Tensor), matching  the definition.

But maybe we need an explicit type somewhere..

**Comment by ckolbPTB** [comment]
I don't have the VS Code setting "mypy-type-checker.importStrategy". Could it be that we are using different plugins?

I am using the mypy plugin from Martan Gover whereas you are probably using the mypy type checker from MS. If I install the mypy type checker from MS then I also get no error messages, although I am not entirely sure it is actually doing anything...

UPDATE:
whereis mypy
mypy: /Users/kolbit01/Documents/PTB/Data/CONDA/MRpro/bin/mypy

mypy -V
mypy 1.4.1 (compiled: yes)

python -V
Python 3.11.4

**Comment by fzimmermann89** [comment]
Something is fishy. I deleted the cache and now I get the same errors in VSCode (which are fixed by #48 ).
Before, VScode canceled the mypy run because of timeouts (still waiting for a new laptop...)
But pre-commit still did not complain. Not sure why tbh.

**Comment by ckolbPTB** [comment]
> Something is fishy.

Agreed!

According to the docu (https://github.com/pre-commit/mirrors-mypy) pre-commit runs `mypy --ignore-missing-imports`

To replicate this, we would have to remove this
```
[[tool.mypy.overrides]]
module = ["ismrmrd.*", "h5py", "scipy.*"]
```
from the pyproject.toml such that `ignore_missing_imports = true` applies to everything. Not sure this is a good idea though. On the other hand different results from VS Code mypy and pre-commit mypy are also not ideal. Any opinions?

It would of course still not fix the above problem with the TypeErrors.

**Comment by fzimmermann89** [comment]
(part 1/3)
I spend way to much time trying to understand this issue, this is what I got so far:
- The pre-commit, as mentioned by @ckolbPTB , runs with `--ignore-missing-imports `. 
-  It also runs in a new virtual environment, without numpy or torch.
-  Thus, whenever it encounters a torch type, it would fail - if not for the ignore-missing-imports. This makes all ` torch.Tensors` in pre-commit `Any`

Test Case:
```python

import torch
def ignored(a: torch.Tensor) -> str:
    reveal_type(a)
    return a


def noticed(a: float) -> str:
    reveal_type(a)
    return a
```
results for pre-commit in 

```
note: Revealed type is "Any"
Revealed type is "builtins.float"
error: Incompatible return value type (got "float", expected "str")  [return-value]
```
--> The torch.Tensor got retyped as any, thus there is no error.

**Comment by fzimmermann89** [comment]
(part 2/3)

We can fix this by installing at least torch and numpy in the pre-commit-mypy environment as well.
```
  - repo: https://github.com/pre-commit/mirrors-mypy
    rev: v1.5.1
    hooks:
      - id: mypy
        additional_dependencies:
          - numpy
          - torch
          - --index-url=https://download.pytorch.org/whl/cpu
          - --extra-index-url=https://pypi.python.org/simple
        exclude: "docs/.*|tests/.*"
```
This now uses 
-  a current version of mypy instead of 0.9
-  installs torch into the mypy venv.

The reasons for the index/extra index lines is to install the cpu-only torch version if run under Linux. Otherwise, a >700MB download has to be done each time the pre-commit venv is created (For example, on each github action run!)

I will open a PR #49  ...  *sigh*

**Comment by fzimmermann89** [comment]
(part 3/3)


> To replicate this, we would have to remove this
> 
> ```
> [[tool.mypy.overrides]]
> module = ["ismrmrd.*", "h5py", "scipy.*"]
> ```
> 
> from the pyproject.toml such that `ignore_missing_imports = true` applies to everything. Not sure this is a good idea though. On the other hand different results from VS Code mypy and pre-commit mypy are also not ideal. Any opinions?

I would prefer to get warned about missing imports by VSCode, i.e. not apply a general `ignore_missing_imports = true`  to everything.
I would also like to get typing errors regarding Tensors, so we could either add all libraries we use for typing to the pre-commit mypy hook, or instead let this hook run in the local venv.

The advantage of running it in the local venv is, that we do not have to keep a separate list if dependencies in the` .pre-commit-config.yaml `. Disadvantage would be, that the local environment influences the errors -> It could be, that one developer sees some errors, another one does not.

So slight preference for the first option. Opinions?

**Comment by schuenke** [comment]
I definitely agree that we should not use a general `ignore_missing_imports = True`.

Currently I think that there won't be to many dependencies that we need to add to the pre-commit config, no?! Thus, I would vote to add these (few) required packages to the config.
